### PR TITLE
Refactor Provider Module and Add Tests

### PR DESCRIPTION
**Pull request by SouzaWander**
Why:

- Less modularity, dependent on Ethereumex, and difficult to test.
- Lack of tests for the provider methods.
- Lack of clarity in the module structure.

How:

- Refactor the Provider module to retrieve environment variables at compile time for improved performance and reliability.
- Optimize the Provider module by using the apply function for Ethereumex method calls.
- Implement test cases to ensure the correctness of the Provider module's methods.

### CODE REVIEW

**Comment by SouzaWander** [comment]
Note: I have been searching, but I did not manage to figure out how to remove the unnecessary argument "def eth_block_number(_args \ [])" from the method's arguments without encountering an error in the mock test. Any suggestions will be highly appreciated. It makes no sense in the code and looks kind of ugly. 😅

File: test/test_helper.exs [hunk 1]
```diff
@@ -1 +1,3 @@
 ExUnit.start()
+Mox.defmock(HttpMock, for: Wrapeth.Provider.Behaviour)
+Application.put_env(:wrapth, ProviderTest, Wrapeth.MockProvider)
```
Comment (joaop21) [inline test/test_helper.exs:3]:
```suggestion
Application.put_env(:wrapeth, ProviderTest, Wrapeth.MockProvider)
```
Besides the typo, why do you need this line?

File: test/wrapeth/provider_test.exs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  # Make sure mocks are verified when the test exits
+  setup :verify_on_exit!
+
+  test "gets accounts" do
+    HttpMock
+    |> expect(:eth_accounts, fn _ -> {:ok, ["0x407d73d8a49eeb85d32cf465507dd71d507100c1"]} end)
+
+    assert TestProvider.eth_accounts() ==
+             ["0x407d73d8a49eeb85d32cf465507dd71d507100c1"]
```
Comment (joaop21) [inline test/wrapeth/provider_test.exs:17]: What about extracting repetitions to another variable? This way, you would only need to set it once.

File: test/wrapeth/provider_test.exs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+  test "gets transaction count" do
+    HttpMock
+    |> expect(:eth_get_transaction_count, fn _some_addr, _, _ -> {:ok, "0x123"} end)
+
+    assert TestProvider.eth_get_transaction_count("some_address", "latest") ==
+             "0x123"
+  end
+
+  test "gets transaction count by nu,ber" do
```
Comment (joaop21) [inline test/wrapeth/provider_test.exs:44]:
```suggestion
  test "gets transaction count by number" do
```

File: test/wrapeth/provider_test.exs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    HttpMock
+    |> expect(:eth_gas_price, fn _ -> {:ok, "0x123"} end)
+
+    assert TestProvider.call_client(:eth_gas_price) ==
+             {:ok, "0x123"}
+  end
+
+  test "error calling client" do
+    HttpMock
+    |> expect(:eth_gas_price, fn _ -> {:error, "error calling the client"} end)
```
Comment (joaop21) [inline test/wrapeth/provider_test.exs:86]: Is this how Ethereumex returns errors?

File: lib/wrapeth/provider/behaviour.ex [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  # API methods
+  @callback eth_accounts(any()) :: any()
+  @callback eth_block_number(any) :: hex_value()
+  @callback eth_get_balance(address, String.t(), any()) :: hex_value()
+  @callback eth_get_block_by_number(hex_value(), boolean(), any()) :: map()
+  @callback eth_get_transaction_count(address(), hex_value(), any()) :: hex_value()
+  @callback eth_get_block_transaction_count_by_number(hex_value(), any()) :: hex_value()
+  @callback eth_gas_price(any()) :: hex_value()
+  @callback eth_get_transaction_receipt(hex_value(), any()) :: map()
+  @callback call_client(keyword(), [String.t() | boolean()], any()) ::
```
Comment (joaop21) [inline lib/wrapeth/provider/behaviour.ex:15]: Why is this function part of the behaviour?
You have to have in mind that future providers will implement this behaviour and they might not implement this `call_client`.

File: lib/wrapeth/provider.ex [hunk 6]
```diff
@@ -1,86 +1,73 @@
...
-        {:ok, receipt} =
-          module_name.eth_get_transaction_receipt(hash, url: node_url)
+      def eth_get_transaction_receipt(hash, _args \\ []) do
+        {:ok, receipt} = call_client(:eth_get_transaction_receipt, [hash])
 
         receipt
       end
+
+      @impl true
+      def call_client(method_name, args \\ [], _arg \\ []) do
```
Comment (joaop21) [inline lib/wrapeth/provider.ex:65]: Should this function be public?

**Comment by SouzaWander** [comment]
@joaop21 I believe I have resolved the issues. Do you think the branch is now ready to be merged? Thanks for your feedback.🙏

File: lib/wrapeth/provider.ex [hunk 7]
```diff
@@ -1,86 +1,72 @@
...
+      def eth_get_transaction_receipt(hash, _args \\ []) do
+        {:ok, receipt} = call_client(:eth_get_transaction_receipt, [hash])
 
         receipt
       end
+
+      defp call_client(method_name, args \\ [], _arg \\ []) do
+        case apply(@client_type, method_name, args ++ [[{:url, @node_url}]]) do
+          {:ok, value} -> {:ok, value}
+          {:error, error} -> raise(error)
```
Comment (joaop21) [inline lib/wrapeth/provider.ex:67]: What do you think it's best here? Raising or returning an error?

File: test/wrapeth/provider_test.exs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  defmodule TestProvider do
+    use Wrapeth.Provider, otp_app: :wrapeth
+  end
+
+  # Make sure mocks are verified when the test exits
+  setup :verify_on_exit!
+
+  @valid_address "0x4Bd040d48fdD1C667fcE5fdDd326681766E6ad91"
+  @invalid_addres "0y4Bd040d48fdD1C667fcE5fdDd326681766E6ad91"
+  @invalid_block_number "0y9710d"
```
Comment (joaop21) [inline test/wrapeth/provider_test.exs:14]: Do you prefer module attributes in tests over setup? Why?

**Comment by joaop21** [comment]
Just a general note. We should have a discussion in threads rather than just resolving it without saying anything. And it's normal for the one who initiates the thread to resolve it. Otherwise, I won't be clarified, and I need to know your thoughts and possible solutions.

File: lib/wrapeth/provider.ex (hunk 7 above)
Comment (SouzaWander) [reply lib/wrapeth/provider.ex:67]: You are right. it's a better idea to return the error. This way, the future user of the tool can have the flexibility to decide how to handle it.

File: test/wrapeth/provider_test.exs (hunk 8 above)
Comment (SouzaWander) [reply test/wrapeth/provider_test.exs]: To be honest, I initially went with the first approach that came to my mind. After looking into the setup functions, I found it to be a more fine-grained approach. I have decided to switch to using the setup function in the last commit.

File: test/wrapeth/provider_test.exs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    {:ok, valid_address: "0x4Bd040d48fdD1C667fcE5fdDd326681766E6ad91"}
+  end
+
+  setup do
+    {:ok, invalid_address: "0y4Bd040d48fdD1C667fcE5fdDd326681766E6ad91"}
+  end
+
+  setup do
+    {:ok, invalid_block_number: "0y9710d"}
+  end
```
Comment (joaop21) [inline test/wrapeth/provider_test.exs:15]: Do you need this to be separated into different setups?

File: lib/wrapeth/provider.ex [hunk 10]
```diff
@@ -1,85 +1,60 @@
...
+      def eth_get_transaction_receipt(hash, _args \\ []) do
+        call_client(:eth_get_transaction_receipt, [hash])
+      end
 
-        receipt
+      defp call_client(method_name, args \\ [], _arg \\ []) do
+        case apply(@client_type, method_name, args ++ [[{:url, @node_url}]]) do
+          {:ok, value} -> {:ok, value}
+          {:error, error} -> {:error, error}
+        end
```
Comment (joaop21) [inline lib/wrapeth/provider.ex:57]: Doesn't this seem weird?
You have 2 cases, and you're retrieving the same thing in both.

Comment (SouzaWander) [reply lib/wrapeth/provider.ex:57]: I just thought that the code would be more readable like this, but, in fact, this case statement is trivial.

File: test/wrapeth/provider_test.exs (hunk 9 above)
Comment (SouzaWander) [reply test/wrapeth/provider_test.exs]: It was a noob move, just fixed in the last commit.

Comment (joaop21) [reply test/wrapeth/provider_test.exs]: Not noob, but I had the idea that ExUnit documentation was more explicit about this.
You can also return this as a Keyword and don't need the tuple.

**Review by joaop21 — APPROVED** [review]

**Pull request merged**